File: web/src/lib/trustTier.ts
/**
 * TrustMeter tier mapping for FE display.
 *
 * NOTE: This is a deterministic stand-in until the listings API includes
 * `seller.trustMeter.tier` in the payload (Phase E). The hash gives every
 * card a stable, varied ribbon based on the sellerId — never random per
 * render, never network-dependent. Replace `deriveTierFromId` with a real
 * lookup against `/api/v1/trust-meter/users/:id` when wiring TrustMeter.
 */
export type TrustTier = 'NEW' | 'BRONZE' | 'SILVER' | 'GOLD' | 'PLATINUM';

const TIER_DISTRIBUTION: readonly TrustTier[] = [
  'NEW', 'NEW', 'NEW',          // 30% — most sellers are new
  'BRONZE', 'BRONZE',           // 20%
  'SILVER', 'SILVER',           // 20%
  'GOLD', 'GOLD',               // 20%
  'PLATINUM',                   // 10% — rare ribbon
] as const;

export function deriveTierFromId(sellerId: string): TrustTier {
  if (!sellerId) return 'NEW';
  let hash = 0;
  for (let i = 0; i < sellerId.length; i++) {
    hash = (hash * 31 + sellerId.charCodeAt(i)) >>> 0;
  }
  return TIER_DISTRIBUTION[hash % TIER_DISTRIBUTION.length];
}

export function tierLabelKey(tier: TrustTier): string {
  switch (tier) {
    case 'NEW':      return 'tierNew';
    case 'BRONZE':   return 'tierBronze';
    case 'SILVER':   return 'tierSilver';
    case 'GOLD':     return 'tierGold';
    case 'PLATINUM': return 'tierPlatinum';
  }
}

export const tierClassMap: Record<TrustTier, string> = {
  NEW:      'tierNew',
  BRONZE:   'tierBronze',
  SILVER:   'tierSilver',
  GOLD:     'tierGold',
  PLATINUM: 'tierPlatinum',
};

File: mobile/lib/core/trust_tier.dart
import 'dart:ui';

/// TrustMeter tier mapping for display.
///
/// Ported from the web app's `src/lib/trustTier.ts`.
/// Deterministic stand-in until the listings API includes
/// `seller.trustMeter.tier` in the payload.
typedef TrustTier = String;

const List<TrustTier> _tierDistribution = [
  'NEW', 'NEW', 'NEW',
  'BRONZE', 'BRONZE',
  'SILVER', 'SILVER',
  'GOLD', 'GOLD',
  'PLATINUM',
];

/// Derives a stable tier from a seller ID using a hash.
/// Never random per render, never network-dependent.
TrustTier deriveTierFromId(String sellerId) {
  if (sellerId.isEmpty) return 'NEW';
  int hash = 0;
  for (int i = 0; i < sellerId.length; i++) {
    hash = (hash * 31 + sellerId.codeUnitAt(i)) & 0xFFFFFFFF;
  }
  return _tierDistribution[hash % _tierDistribution.length];
}

/// Returns the color for a trust tier.
int tierColor(TrustTier tier) {
  switch (tier) {
    case 'NEW':
      return 0xFF94A3B8;
    case 'BRONZE':
      return 0xFFB45309;
    case 'SILVER':
      return 0xFF64748B;
    case 'GOLD':
      return 0xFFCA8A04;
    case 'PLATINUM':
      return 0xFF0E7490;
    default:
      return 0xFF94A3B8;
  }
}

/// Returns the tier color as a Color object.
Color tierColorValue(TrustTier tier) => Color(tierColor(tier));

/// Returns the i18n key suffix for a tier.
String tierLabelKey(TrustTier tier) {
  switch (tier) {
    case 'NEW':
      return 'listingTierNew';
    case 'BRONZE':
      return 'listingTierBronze';
    case 'SILVER':
      return 'listingTierSilver';
    case 'GOLD':
      return 'listingTierGold';
    case 'PLATINUM':
      return 'listingTierPlatinum';
    default:
      return 'listingTierNew';
  }
}
